luno_python/stream_client.py

```diff
@@ -0,0 +1,201 @@
+    order = orders.pop(order_id, None)
+    if order is None:
+        return
+
+    new_order = order._replace(volume=order.volume - volume)
+    if new_order.volume > DEC_0:
+        orders[order_id] = new_order
+
+
+class _MarketStreamState:
+    def __init__(self, first: dict):
+        if first is None:
+            raise Exception("Unable to use empty message to initialise market state")
+
+        def conv_message(msg):
+            return Order(
+                msg['id'],
+                Decimal(msg['price']),
+                Decimal(msg['volume']),
+            )
+
+        bids = [conv_message(m) for m in first['bids']]
+        asks = [conv_message(m) for m in first['asks']]
+        self._bids = {b.order_id: b for b in bids}
+        self._asks = {a.order_id: a for a in asks}
+        self._sequence = first['sequence']
+        self._trades = []
+        self._status = first['status']
+
+    def get_asks(self):
+        return _flatten_orders(self._asks, False)
+
+    def get_bids(self):
+        return _flatten_orders(self._bids, True)
+
+    def get_status(self):
+        return self._status
+
+    def get_sequence(self):
+        return self._sequence
+
+    def get_snapshot(self):
+        return MarketState(
+            sequence=self.get_sequence(),
+            asks=self.get_asks(),
+            bids=self.get_bids(),
+            status=self.get_status(),
+        )
+
+    def process_update(self, update: dict):
+        if update is None:
+            return
+
+        seq = update['sequence']
+        if int(seq) != int(self._sequence)+1:
+            raise OutOfOrderMessageException()
+
+        trades = update.get('trade_updates')
+        if trades:
+            self._process_trades(trades)
+
+        create = update.get('create_update')
+        if create:
+            self._process_create(create)
+
+        delete_upd = update.get('delete_update')
+        if delete_upd:
+            self._process_delete(delete_upd)
+
+        status_upd = update.get('status_update')
+        if status_upd:
+            self._process_status(status_upd)
+
+        self._sequence = seq
+
+    def _process_trades(self, trade_updates: List[dict]):
+        for t in trade_updates:
+            maker_id = t['maker_order_id']
+            volume = Decimal(t['base'])
+
+            _decrement_trade(self._asks, maker_id, volume)
+            _decrement_trade(self._bids, maker_id, volume)
+
+    def _process_create(self, create_update: dict):
+        o = Order(
+            create_update['order_id'],
+            Decimal(create_update['price']),
+            Decimal(create_update['volume']),
+        )
+        if create_update['type'] == "ASK":
+            self._asks[o.order_id] = o
+        elif create_update['type'] == "BID":
+            self._bids[o.order_id] = o
+
+    def _process_delete(self, delete_update: dict):
+        order_id = delete_update['order_id']
+        self._asks.pop(order_id, None)
```

Comment:
Why not `self._asks.pop(order_id)` ? Raising a `KeyError` if `order_id` is not found might be preferred.